### PR TITLE
[Feature] New pipeline & executor for SDK

### PR DESCRIPTION
**Pull request by lzhangzz**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: csrc/apis/c/executor.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
+//
+// Created by zhangli on 4/26/22.
```
Comment (lvhan028) [inline csrc/apis/c/executor.h:2]: License issue

File: csrc/execution/schedulers/registry.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+//
+// Created by zhangli on 4/26/22.
```
Comment (lvhan028) [inline csrc/execution/schedulers/registry.h:2]: license issue

File: tests/test_csrc/capi/test_detector.cpp [hunk 3]
```diff
@@ -64,3 +64,32 @@ TEST_CASE("test detector's c api", "[detector]") {
     }
   }
 }
+
+#if 0
```
Comment (lvhan028) [inline tests/test_csrc/capi/test_detector.cpp:68]: Why "#if 0"

**Comment by lvhan028** [comment]
try https://github.com/marketplace/actions/clang-format-check to fix linting issues.

File: csrc/apis/c/common.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  return Guard([&] { return Take(Value(*Cast(input))); });
+}
+
+int mmdeploy_value_destroy(mmdeploy_value_t value) {
+  delete Cast(value);
+  return 0;
+}
+
+int mmdeploy_common_create_input(const mm_mat_t* mats, int mat_count, mmdeploy_value_t* value) {
+  if (mat_count && mats == nullptr) {
```
Comment (lvhan028) [inline csrc/apis/c/common.cpp:20]: Is `mat_count == 0` a valid input?

**Comment by irexyc** [comment]
Can't build on windows

**Comment by grimoire** [comment]
cmake failed with log:

```
CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
  "mmdeploy_executor" of type SHARED_LIBRARY
    depends on "mmdeploy_common" (weak)
    depends on "mmdeploy_model" (weak)
    depends on "mmdeploy_pipeline" (weak)
  "mmdeploy_model" of type SHARED_LIBRARY
    depends on "mmdeploy_common" (weak)
    depends on "mmdeploy_executor" (weak)
    depends on "mmdeploy_pipeline" (weak)
  "mmdeploy_pipeline" of type SHARED_LIBRARY
    depends on "mmdeploy_common" (weak)
    depends on "mmdeploy_executor" (weak)
    depends on "mmdeploy_model" (weak)
  "mmdeploy_common" of type SHARED_LIBRARY
    depends on "mmdeploy_executor" (weak)
    depends on "mmdeploy_model" (weak)
    depends on "mmdeploy_pipeline" (weak)
At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are allowed only among static libraries.

```

File: csrc/apis/c/CMakeLists.txt [hunk 5]
```diff
@@ -4,38 +4,62 @@ project(capis)
...
         list(APPEND TASK_LIST "rotated_detector")
-    endif()
+    endif ()
 endif ()
 
 foreach (TASK ${TASK_LIST})
     set(TARGET_NAME mmdeploy_${TASK})
-    mmdeploy_add_library(${TARGET_NAME} ${TASK}.cpp)
+    mmdeploy_add_library(${TARGET_NAME} ${CMAKE_CURRENT_SOURCE_DIR}/${TASK}.cpp)
+    if (NOT TASK STREQUAL common)
```
Comment (lvhan028) [inline csrc/apis/c/CMakeLists.txt:51]: Cyclic dependencies among `mmdeploy_common`, `mmdeploy_executor`, `mmdeploy_model` and `mmdeploy_pipline` 
I think `${TARGET_NAME}` is trying to link `mmdeploy_common`, `mmdeploy_executor`, `mmdeploy_model` and `mmdeploy_pipline`, but the logic is not correct

**Comment by lvhan028** [comment]
@irexyc @grimoire please build static lib with `-DBUILD_SHARED_LIBS=OFF` at first, and keep on reviewing.

**Comment by lzhangzz** [comment]
dynamic libraries should be working now.

File: csrc/apis/c/common.cpp (hunk 4 above)
Comment (lzhangzz) [reply csrc/apis/c/common.cpp:20]: yes

File: csrc/execution/type_traits.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  using type = T;
+};
+
+template <typename T>
+using remove_cvref_t = typename remove_cvref<T>::type;
+
+template <typename Fn, typename... Args>
+using callable_result_t = decltype(std::declval<Fn&&>()(std::declval<Args&&>()...));
+
+namespace _is_callable {
```
Comment (grimoire) [inline csrc/execution/type_traits.h:45]: It seem that `is_callable` and `defer` trait has not been used. Did you plan to use it in `tag_invoke`?

File: csrc/execution/tag_invoke.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+template <typename CPO, typename... Args>
+inline constexpr bool is_nothrow_tag_invocable_v =
+    noexcept(_tag_invoke::try_tag_invoke<CPO, Args...>(0));
+
+template <typename CPO, typename... Args>
+using is_nothrow_tag_invocable = std::bool_constant<is_nothrow_tag_invocable_v<CPO, Args...>>;
+
+template <typename CPO, typename... Args>
+inline constexpr bool tag_invocable = (sizeof(_tag_invoke::try_tag_invoke<CPO, Args...>(0)) ==
```
Comment (grimoire) [inline csrc/execution/tag_invoke.h:78]: Is `tag_invocable` the same as `is_tag_invocable_v`?

File: csrc/apis/c/executor.h [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+MMDEPLOY_API mmdeploy_sender_t mmdeploy_executor_just(mmdeploy_value_t value);
+
+/**
+ * @brief
+ * @param[in] scheduler
+ * @return the sender created
+ */
+MMDEPLOY_API mmdeploy_sender_t mmdeploy_executor_schedule(mmdeploy_scheduler_t scheduler);
+
+MMDEPLOY_API mmdeploy_sender_t mmdeploy_executor_transfer_just(mmdeploy_scheduler_t scheduler,
```
Comment (grimoire) [inline csrc/apis/c/executor.h:94]: I can't find the definition of this function.

**Comment by lzhangzz** [comment]
> `split` failed after `when all`.
> 
> ```c++
>   Value v(1);
>   mmdeploy_value_t value = Cast(&v);
>   auto just = mmdeploy_executor_just(value);
>   auto split = mmdeploy_executor_split(just);
>   auto then1 = mmdeploy_executor_then(
>       split,
>       [](mmdeploy_value_t value, void* context) {
>         std::cout << "split then1\n";
>         return value;
>       },
>       nullptr);
>   auto then2 = mmdeploy_executor_then(
>       split,
>       [](mmdeploy_value_t value, void* context) {
>         std::cout << "split then2\n";
>         return value;
>       },
>       nullptr);
> 
>   mmdeploy_sender_t when_all_input[] = {then1, then2};
>   auto when_all = mmdeploy_executor_when_all(when_all_input, 2);
> 
>   auto result = mmdeploy_executor_sync_wait(when_all);
> ```
> 
> error log:
> 
> ```
> free(): double free detected in tcache 2
> Aborted (core dumped)
> ```
> 
> It is ok if I connect two different `just` to `then`. should `split` be multi shot?

updated docs for `mmdeploy_executor_split`

All sender APIs are destructive, they consume there input sender. This design saves the users from destroying every sender they encounter. `mmdeploy_executor_split` just make it's returned sender copyable so that it can be copied multiple times.

```C++
  Value v(1);
  mmdeploy_value_t value = Cast(&v);
  auto just = mmdeploy_executor_just(value);
  auto split1 = mmdeploy_executor_split(just);
  auto split2 = mmdeploy_sender_copy(split1);
  auto then1 = mmdeploy_executor_then(
      split1,
      [](mmdeploy_value_t value, void* context) {
        std::cout << "split then1\n";
        return value;
      },
      nullptr);
  auto then2 = mmdeploy_executor_then(
      split2,
      [](mmdeploy_value_t value, void* context) {
        std::cout << "split then2\n";
        return value;
      },
      nullptr);

  mmdeploy_sender_t when_all_input[] = {then1, then2};
  auto when_all = mmdeploy_executor_when_all(when_all_input, 2);

  auto result = mmdeploy_executor_sync_wait(when_all);
```

**Comment by lzhangzz** [comment]
@irexyc MSVC shared libs build should be working now

**Review by grimoire — APPROVED** [review]

**Review by irexyc — APPROVED** [review]

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**